File: src/generators/ipv4_iter.rs
use std::net::Ipv4Addr;
use crate::utils::abort;



#[derive(Clone)]
pub(crate) struct Ipv4Iter {
    current : u32,
    end     : u32,
    total   : u64,
    pub start_u32 : u32,
    pub end_u32   : u32,
}


#[derive(Default)]
struct TempData {
    range               : String,
    cidr_has_usable_ips : bool,
    usable_start        : u32,
    usable_end          : u32,
    start_part          : String,
    end_part            : String,
    start_ip            : Option<u32>,
    end_ip              : Option<u32>,
    start_in_cidr       : bool,
    end_in_cidr         : bool,
}


impl Ipv4Iter {
    
    pub fn new(cidr: &str, range: Option<&str>) -> Self {
        let mut data = TempData::default();

        let (network_u32, broadcast_u32) = Self::parse_cidr(cidr);
        
        data.usable_start        = network_u32.saturating_add(1);
        data.usable_end          = broadcast_u32.saturating_sub(1);
        data.cidr_has_usable_ips = data.usable_start <= data.usable_end;
        
        let (start_range, end_range) = if let Some(range_str) = range {
            data.range = range_str.trim().to_string();
            Self::parse_range(&mut data)
        } else {
            if data.cidr_has_usable_ips {
                (data.usable_start, data.usable_end)
            } else {
                (network_u32, network_u32)
            }
        };

        if start_range > end_range {
            abort("Start IP cannot be greater than end IP");
        }

        let total = (end_range - start_range + 1) as u64;

        Ipv4Iter {
            current   : start_range,
            end       : end_range,
            start_u32 : start_range,
            end_u32   : end_range,
            total,
        }
    }



    fn parse_cidr(cidr: &str) -> (u32, u32) {
        let parts: Vec<&str> = cidr.split('/').collect();
        if parts.len() != 2 {
            abort(&format!("Invalid CIDR: {}", cidr));
        }

        let ip: Ipv4Addr = parts[0].parse()
            .unwrap_or_else(|e| {
                abort(&format!("Invalid IP in CIDR '{}': {}", cidr, e));
        });

        let prefix: u8 = parts[1].parse::<u8>()
            .unwrap_or_else(|e| {
                abort(&format!("Invalid prefix in CIDR '{}': {}", cidr, e));
        });
        
        if prefix > 32 {
            abort(&format!("Prefix out of range in CIDR '{}': {}", cidr, prefix));
        }

        let ip_u32 = u32::from_be_bytes(ip.octets());
        let network_mask = if prefix == 0 {
            0u32
        } else {
            (!0u32).checked_shl(32 - prefix as u32).unwrap_or(0)
        };
        
        let network_u32   = ip_u32 & network_mask;
        let broadcast_u32 = network_u32 | !network_mask;

        (network_u32, broadcast_u32)
    }



    fn parse_range(data: &mut TempData) -> (u32, u32) {
        if data.range.is_empty() {
            abort("Range string cannot be empty");
        }

        if data.range.contains('*') {
            Self::parse_wildcard_range(data)
        } else {
            Self::parse_single_ip_range(data)
        }
    }



    fn parse_single_ip_range(data: &TempData) -> (u32, u32) {
        let ip     = Self::parse_ip_address(&data.range);
        let ip_u32 = u32::from_be_bytes(ip.octets());
        (ip_u32, ip_u32)
    }

    
    
    fn parse_wildcard_range(data: &mut TempData) -> (u32, u32) {
        let (start_part, end_part) = Self::split_wildcard_parts(data);
        data.start_part            = start_part.clone();
        data.end_part              = end_part.clone();

        let (start_ip, start_in_cidr) = Self::parse_ip_part(data, &start_part);
        data.start_ip                 = start_ip;
        data.start_in_cidr            = start_in_cidr;

        let (end_ip, end_in_cidr) = Self::parse_ip_part(data, &end_part);
        data.end_ip               = end_ip;
        data.end_in_cidr          = end_in_cidr;

        Self::determine_range_bounds(data)
    }

    
    
    fn split_wildcard_parts(data: &TempData) -> (String, String) {
        let parts: Vec<&str> = data.range.split('*').collect();

        if parts.len() != 2 {
            abort(&format!("Invalid range format: {}", data.range));
        }

        (parts[0].trim().to_string(), parts[1].trim().to_string())
    }

    
    
    fn parse_ip_part(data: &TempData, ip_str: &str) -> (Option<u32>, bool) {
        if ip_str.is_empty() {
            return (None, false);
        }

        let ip      = Self::parse_ip_address(ip_str);
        let ip_u32  = u32::from_be_bytes(ip.octets());
        let in_cidr = 
            data.cidr_has_usable_ips && 
            ip_u32 >= data.usable_start && 
            ip_u32 <= data.usable_end;

        (Some(ip_u32), in_cidr)
    }

    
    
    fn parse_ip_address(ip_str: &str) -> Ipv4Addr {
        ip_str.parse().unwrap_or_else(|e| {
            abort(&format!("Invalid IP address '{}': {}", ip_str, e));
        })
    }



    fn determine_range_bounds(data: &mut TempData) -> (u32, u32) {
        match (data.start_part.is_empty(), data.end_part.is_empty()) {
            (false, false) => { Self::get_start_and_end_ip(data) },
            (false, true)  => { Self::get_start_ip_and_usable_end(data) },
            (true,  false) => { Self::get_usable_start_and_end_ip(data) },
            (true,  true)  => { Self::get_usable_start_and_end(data) },
        }
    }



    fn get_start_and_end_ip(data: &TempData) -> (u32, u32) {
        let start_ip = data.start_ip.clone();
        let end_ip   = data.end_ip.clone();

        (start_ip.unwrap(), end_ip.unwrap())
    }



    fn get_start_ip_and_usable_end(data: &mut TempData) -> (u32, u32) {
        let ip = data.start_ip.clone();

        if !data.start_in_cidr {
            let ip_u32 = ip.unwrap();
            
            abort(&format!(
                "Start IP {} is outside CIDR range. When using 'IP*', the IP must be within the CIDR",
                Ipv4Addr::from(ip_u32.to_be_bytes())
            ));
        }
        
        (ip.unwrap(), data.usable_end.clone())
    }



    fn get_usable_start_and_end_ip(data: &TempData) -> (u32, u32) {
        let ip = data.end_ip.clone();

        if !data.end_in_cidr {
            let ip_u32 = ip.unwrap();
            
            abort(&format!(
                "End IP {} is outside CIDR range. When using '*IP', the IP must be within the CIDR.",
                Ipv4Addr::from(ip_u32.to_be_bytes())
            ));
        }
        
        (data.usable_start.clone(), ip.unwrap())
    }



    fn get_usable_start_and_end(data: &TempData) -> (u32, u32) {
        if data.cidr_has_usable_ips {
            return (data.usable_start.clone(), data.usable_end.clone());
        }
        
        (data.usable_start.clone(), data.usable_start.clone())
    }



    pub fn total(&self) -> u64 {
        self.total
    }

}



impl Iterator for Ipv4Iter {
    type Item = Ipv4Addr;

    fn next(&mut self) -> Option<Ipv4Addr> {
        if self.current > self.end {
            return None;
        }
            
        let ip = Ipv4Addr::from(self.current.to_be_bytes());
        self.current += 1;
        Some(ip)
    }
}

File: generators/ipv4_iter.go
/*
 * Copyright (C) 2025 Oliver R. Calazans Jeronimo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org>.
 */

package generators

import (
	"fmt"
	"net"
	"offscan/conv"
	"offscan/utils"
	"strconv"
	"strings"
)



type Ipv4Iter struct {
    current  uint32
    end      uint32
    total    uint64
    StartU32 uint32
    EndU32   uint32
}



func NewIpv4Iter(cidr string, rangeStr *string) *Ipv4Iter {
    networkU32, broadcastU32 := parseCIDR(cidr)

    usableStart   := networkU32 + 1
    usableEnd     := broadcastU32 - 1
    cidrHasUsable := usableStart <= usableEnd

    var startRange, endRange uint32
    
    if rangeStr != nil && *rangeStr != "" {
        s := strings.TrimSpace(*rangeStr)
        startRange, endRange  = parseRange(s, usableStart, usableEnd, cidrHasUsable)
    
	} else {
        if cidrHasUsable {
            startRange, endRange = usableStart, usableEnd
        } else {
            startRange, endRange = networkU32, networkU32
        }
    }

    if startRange > endRange {
        utils.Abort("Start IP cannot be greater than end IP")
    }

    total := uint64(endRange - startRange + 1)
    
    return &Ipv4Iter{
        current:  startRange,
        end:      endRange,
        total:    total,
        StartU32: startRange,
        EndU32:   endRange,
    }
}



func parseCIDR(cidr string) (network, broadcast uint32) {
    parts := strings.Split(cidr, "/")
    
	if len(parts) != 2 {
        utils.Abort(fmt.Sprintf("Invalid CIDR: %s", cidr))
    }

    ip := net.ParseIP(parts[0])
    if ip == nil {
        utils.Abort(fmt.Sprintf("Invalid IP in CIDR '%s'", cidr))
    }
    
	ip = ip.To4()
    if ip == nil {
        utils.Abort(fmt.Sprintf("CIDR '%s' is not IPv4", cidr))
    }

    prefix, err := strconv.Atoi(parts[1])
    if err != nil || prefix < 0 || prefix > 32 {
        utils.Abort(fmt.Sprintf("Invalid prefix in CIDR '%s'", cidr))
    }

    ipU32 := conv.IPToU32(ip)

    var mask uint32
    if prefix == 0 {
        mask = 0
    } else {
        mask = ^uint32(0) << (32 - uint(prefix))
    }

    network   = ipU32 & mask
    broadcast = network | ^mask
    
	return network, broadcast
}



func parseRange(
	r 			  string,  
	usableStart   uint32, 
	usableEnd     uint32, 
	cidrHasUsable bool,
) (uint32, uint32) {
    if strings.Contains(r, "*") {
        return parseWildcardRange(r, usableStart, usableEnd, cidrHasUsable)
    }

	return parseSingleIPRange(r)
}



func parseSingleIPRange(ipStr string) (uint32, uint32) {
    ip    := parseIPAddress(ipStr)
	ipU32 := conv.IPToU32(ip)
    return ipU32, ipU32
}



func parseWildcardRange(
	rangeStr      string, 
	usableStart   uint32, 
	usableEnd     uint32, 
	cidrHasUsable bool,
) (uint32, uint32) {
	parts := strings.SplitN(rangeStr, "*", 2)
    
	if len(parts) != 2 {
        utils.Abort(fmt.Sprintf("Invalid range format: %s", rangeStr))
    }

	startPart := strings.TrimSpace(parts[0])
    endPart   := strings.TrimSpace(parts[1])

    var startIP *uint32
    var startInCidr bool

	if startPart != "" {
        ip         := parseIPAddress(startPart)
		ipU32      := conv.IPToU32(ip)
        startIP     = &ipU32
        startInCidr = cidrHasUsable && ipU32 >= usableStart && ipU32 <= usableEnd
    }

	var endIP *uint32
    var endInCidr bool

	if endPart != "" {
        ip       := parseIPAddress(endPart)
		ipU32 	 := conv.IPToU32(ip)
        endIP     = &ipU32
        endInCidr = cidrHasUsable && ipU32 >= usableStart && ipU32 <= usableEnd
    }

    switch {
    case startPart != "" && endPart != "":
        return *startIP, *endIP

    case startPart != "" && endPart == "":
        if !startInCidr {
    		ip := conv.U32ToIP(*startIP)
		    utils.Abort(fmt.Sprintf("Start IP %s is outside CIDR range. When using 'IP*', the IP must be within the CIDR", ip.String()))
		}
        
		return *startIP, usableEnd

    case startPart == "" && endPart != "":
        if !endInCidr {
            ip := conv.U32ToIP(*endIP)
            utils.Abort(
                fmt.Sprintf("End IP %s is outside CIDR range. When using '*IP', the IP must be within the CIDR", ip.String()),
            )
        }
        
		return usableStart, *endIP

    case startPart == "" && endPart == "":
        if cidrHasUsable {
            return usableStart, usableEnd
        }
        return usableStart, usableStart

    default:
        utils.Abort("Unexpected wildcard parsing")
        return 0, 0
    }
}



func parseIPAddress(ipStr string) net.IP {
    ip := net.ParseIP(ipStr)
    if ip == nil {
        utils.Abort(fmt.Sprintf("Invalid IP address '%s'", ipStr))
    }

    ip4 := ip.To4()
    if ip4 == nil {
        utils.Abort(fmt.Sprintf("'%s' is not an IPv4 address", ipStr))
    }

    return ip4
}



func (it *Ipv4Iter) Next() (net.IP, bool) {
    if it.current > it.end {
        return nil, false
    }

    ip := conv.U32ToIP(it.current)
    it.current++
    return ip, true
}



func (it *Ipv4Iter) Total() uint64 {
    return it.total
}



func (it *Ipv4Iter) Reset() {
    it.current = it.StartU32
}